File: backend/app/schemas/analytics.py
"""Analytics response schemas matching the frontend chart data contract."""

from typing import List

from pydantic import ConfigDict, Field

from app.schemas.base import CamelModel


class SeasonRange(CamelModel):
    min: int
    max: int


class ConstructorInfo(CamelModel):
    ref: str
    name: str


class ConstructorPoint(CamelModel):
    season: int
    model_config = ConfigDict(
        alias_generator=None,
        populate_by_name=True,
        from_attributes=True,
        extra="allow",
    )
    __pydantic_extra__: dict[str, float] = Field(init=False)


class ConstructorDominance(CamelModel):
    constructors: List[ConstructorInfo]
    points: List[ConstructorPoint]


class PoleLeaderboardRow(CamelModel):
    driver_code: str
    driver_name: str
    poles: int


class FastestLapLeaderboardRow(CamelModel):
    driver_code: str
    driver_name: str
    fastest_laps: int


class AvgQualifyingPoint(CamelModel):
    season: int
    model_config = ConfigDict(
        alias_generator=None,
        populate_by_name=True,
        from_attributes=True,
        extra="allow",
    )
    __pydantic_extra__: dict[str, float] = Field(init=False)


class AvgQualifying(CamelModel):
    driver_codes: List[str]
    points: List[AvgQualifyingPoint]


class PodiumTrendPoint(CamelModel):
    season: int
    model_config = ConfigDict(
        alias_generator=None,
        populate_by_name=True,
        from_attributes=True,
        extra="allow",
    )
    __pydantic_extra__: dict[str, int] = Field(init=False)


class PodiumTrends(CamelModel):
    driver_codes: List[str]
    points: List[PodiumTrendPoint]


class AnalyticsData(CamelModel):
    season_range: SeasonRange
    constructor_dominance: ConstructorDominance
    pole_leaderboard: List[PoleLeaderboardRow]
    fastest_lap_leaderboard: List[FastestLapLeaderboardRow]
    avg_qualifying: AvgQualifying
    podium_trends: PodiumTrends

File: frontend/src/types/analytics.ts
export interface SeasonRange {
  min: number;
  max: number;
}

/** One row per season, one numeric field per constructor's points that season. */
export interface ConstructorDominancePoint {
  season: number;
  [constructorRef: string]: number | string;
}

export interface ConstructorMeta {
  ref: string;
  name: string;
}

export interface PoleLeaderboardRow {
  driverCode: string;
  driverName: string;
  poles: number;
}

export interface FastestLapLeaderboardRow {
  driverCode: string;
  driverName: string;
  fastestLaps: number;
}

/** One row per season, one numeric field per driver's average qualifying position that season. */
export interface AvgQualifyingPoint {
  season: number;
  [driverCode: string]: number | string;
}

/** One row per season, one numeric field per driver's podium count that season. */
export interface PodiumTrendPoint {
  season: number;
  [driverCode: string]: number | string;
}

export interface DriverTeamConstructor {
  ref: string;
  name: string;
  color: string;
}

export interface DriverTeamPoint {
  season: number;
  driverCode: string;
  constructors: DriverTeamConstructor[];
}

export interface AnalyticsData {
  seasonRange: SeasonRange;
  constructorDominance: {
    points: ConstructorDominancePoint[];
    constructors: ConstructorMeta[];
  };
  poleLeaderboard: PoleLeaderboardRow[];
  fastestLapLeaderboard: FastestLapLeaderboardRow[];
  avgQualifying: {
    points: AvgQualifyingPoint[];
    driverCodes: string[];
  };
  podiumTrends: {
    points: PodiumTrendPoint[];
    driverCodes: string[];
  };
  driverTeams: DriverTeamPoint[];
}
